Add READMEs to every package

dmathieu opened the pull request:
Replacement for #5102.

pellared commented on .github/workflows/scripts/readme.sh:
```diff
@@ -0,0 +1,19 @@
+# Copyright The OpenTelemetry Authors
+# SPDX-License-Identifier: Apache-2.0
+
+dirs=(`find . -type f -name "go.mod" -not -path "*/internal/*" -exec dirname {} \; | sort | egrep  '^./'`)
```
This finds non-internal Go modules which is a subset of all non-internal Go packages. For instance https://pkg.go.dev/go.opentelemetry.io/otel/metric/noop is also a package.

dmathieu replied:
Hmm, I thought we'd want to do this for modules, not every package. That seems a tad too much?

pellared replied:
For a regular developer browsing GitHub there is not a big difference between a package and a module.

dmathieu replied:
I don't mind adding it, though it adds quite a lot of readme files.
But the main difference between both may be that it's more likely that folks who just get started with the library would want to use modules than each package?

dmathieu replied:
I've done it. One issue there is that now, we always have issues with the links checker, as we reach the `pkg.go.dev` requests limit.

pellared replied:
AFAIK it is possible to configure `lychee` to ignore `429` errors.

I also see that the specification repo uses a different action for checking hyperlinks and retries on `429`. See:
- https://github.com/open-telemetry/opentelemetry-specification/blob/65875014e15fa9739501413b3349961d503c0630/.github/workflows/checks.yaml#L36-L46
- https://github.com/open-telemetry/opentelemetry-specification/blob/65875014e15fa9739501413b3349961d503c0630/.markdown_link_check_config.json#L17

pellared replied:
Related issue https://github.com/lycheeverse/lychee/issues/989

dmathieu replied:
Going to a very low max concurrency of 5 seems to have fixed it for now.

pellared commented on .github/workflows/scripts/readme.sh:
```diff
@@ -0,0 +1,19 @@
+# Copyright The OpenTelemetry Authors
+# SPDX-License-Identifier: Apache-2.0
+
+dirs=(`find . -type f -name "go.mod" -not -path "*/internal/*" -exec dirname {} \; | sort | egrep  '^./'`)
```
How about adding a mention in https://github.com/open-telemetry/opentelemetry-go/blob/main/CONTRIBUTING.md#documentation that each Go modules should have a `README.md` with a title and PkgGoDev badge?

pellared replied:
How about ignoring directories that are named `test`?

pellared replied:
How about ignoring `*/example/*`?

pellared reviewed (commented):
I like this 👍

MrAlias commented:
Are we moving away from package being is go files? I thought the group decided to not straddle having it there and in a README.md?

pellared commented:
> Are we moving away from package being is go files? I thought the group decided to not straddle having it there and in a README.md?

The proposal is to have readme files that would just refer to pkg.go.dev. 
See:
- https://github.com/open-telemetry/opentelemetry-go/pull/5102#issuecomment-2017920839
- https://github.com/open-telemetry/opentelemetry-go/pull/5102#issuecomment-2018029858

MrAlias commented:
> > Are we moving away from package being is go files? I thought the group decided to not straddle having it there and in a README.md?
> 
> The proposal is to have readme files that would just refer to pkg.go.dev. See:
> 
> * [Add a readme file to opencensus #5102 (comment)](https://github.com/open-telemetry/opentelemetry-go/pull/5102#issuecomment-2017920839)
> * [Add a readme file to opencensus #5102 (comment)](https://github.com/open-telemetry/opentelemetry-go/pull/5102#issuecomment-2018029858)

Can we clarify in our contributing guidelines that we do not add documentation to the readme and Go docs?

dmathieu commented:
> Can we clarify in our contributing guidelines that we do not add documentation to the readme and Go docs?

Sure. How about the current wording?

MrAlias commented on .github/workflows/scripts/readme.sh:
```diff
@@ -0,0 +1,19 @@
+	pwd
+	if [ ! -f "README.md" ]; then
+		echo "couldn't find README.md for $dir"
+		exit 1
```
It would be more helpful to have this echo all the directories without a README before it exited. Maybe save this value as a var and exit after the loop?

MrAlias commented on .github/workflows/scripts/readme.sh:
Is there a way to integrate this into the Makefile? Ideally devs can check their PR without having to push to github.

dmathieu commented:
I have moved the check to a script executable from the Makefile, and added it to `make precommit`.
The script will also now run through each package, output for any that fails, and exit with an error if any failed.

```bash
open-telemetry/opentelemetry-go›  git:(readme-for-all-packages) 3D2M 1A make verify-readmes
./verify_readmes.sh
couldn't find README.md for ./bridge/opencensus
couldn't find README.md for ./bridge/opentracing
Error: some READMEs couldn't be found.
make: *** [verify-readmes] Error 1
```

pellared commented on Makefile:
```diff
@@ -8,13 +8,15 @@ ALL_GO_MOD_DIRS := $(shell find . -type f -name 'go.mod' -exec dirname {} \; | s
 OTEL_GO_MOD_DIRS := $(filter-out $(TOOLS_MOD_DIR), $(ALL_GO_MOD_DIRS))
 ALL_COVERAGE_MOD_DIRS := $(shell find . -type f -name 'go.mod' -exec dirname {} \; | grep -E -v '^./example|^$(TOOLS_MOD_DIR)' | sort)
 
+ALL_PACKAGES := $(shell find . -type d -not -path "*/internal*" -not -path "*/test*" -not -path "*/example*" -not -path "*/.*" | sort)
```
Unused?

dmathieu replied:
🤦 yeah, I forgot to remove it.

pellared commented on .github/workflows/readme.yml:
```diff
@@ -0,0 +1,15 @@
+name: ci
```
I think you have forgotten to rename.

Anyway, I propose to add `verify-readmes` target to `ci` in `Makfile` and get rid of this GitHub workflow.

pellared commented on Makefile:
```diff
@@ -14,7 +14,7 @@ TIMEOUT = 60
 .PHONY: precommit ci
-precommit: generate dependabot-generate license-check misspell go-mod-tidy golangci-lint-fix test-default
+precommit: generate dependabot-generate license-check misspell go-mod-tidy golangci-lint-fix verify-readmes test-default
 ci: generate dependabot-check license-check lint vanity-import-check build test-default check-clean-work-tree test-coverage
```
```suggestion
ci: generate dependabot-check license-check lint vanity-import-check verify-readmes build test-default check-clean-work-tree test-coverage
```

pellared reviewed (approved):
🏅

pellared commented on CONTRIBUTING.md:
```diff
@@ -201,6 +201,16 @@ You can install and run a "local Go Doc site" in the following way:
 
+### README files
+
+Each (non-internal, non-test, non-documentation) module must contain a
```
```suggestion
Each (non-internal, non-test, non-documentation) package must contain a
```
